orchestrator/src/server.ts

```diff
@@ -89,30 +95,64 @@ async function runPreprocessorsParallel(data: Record<string, unknown>, preproces
-            fetch(`http://${preprocessor[0]}:${preprocessor[1]}/preprocessor`, {
-                "method": "POST",
-                "headers": {
-                    "Content-Type": "application/json"
-                },
-                "body": JSON.stringify(data),
-                "signal": controller.signal
-            }).then(r => {
-                clearTimeout(timeout);
-                resolve(r);
-            }).catch(err => {
-                console.error("Error occured fetching from " + preprocessor[0]);
-                console.error(err);
-                resolve();
+            // get value from cache for each preprocessor if it exists
+            const cacheTimeOut = preprocessor[3] as number;
+            const reqCapabilities = data["capabilities"] as string[];
+            const isDebugMode = reqCapabilities && reqCapabilities.includes("ca.mcgill.a11y.image.capability.DebugMode")
+            const preprocessorName = SERVICE_PREPROCESSOR_MAP[preprocessor[0]] || '';
+            const cacheKeyData = {"imageBlob": data["graphic"], "preprocessor":preprocessorName, "debugMode":isDebugMode};
+            const hashedKey = hash(cacheKeyData);
+            getResponseFromCache(hashedKey).then((cacheValue)=>{
+                if(cacheTimeOut && cacheValue){
+                    // Return the value from cache if found
+                    console.debug(`Response for preprocessor ${preprocessorName} served from cache`);
+                    const cacheResponse = JSON.parse(cacheValue) as Response;
+                    (data["preprocessors"] as Record<string, unknown>)[preprocessorName] = cacheResponse;
+                    resolve(cacheResponse);
+                } else {
+                    // Call the preprocessor endpoint to get response
+                    console.debug("Sending to preprocessor \"" + preprocessor[0] + "\"");
+                    fetch(`http://${preprocessor[0]}:${preprocessor[1]}/preprocessor`, {
+                        "method": "POST",
+                        "headers": {
+                            "Content-Type": "application/json"
+                        },
+                        "body": JSON.stringify(data),
+                        "signal": controller.signal
+                    }).then(r => {
+                        clearTimeout(timeout);
+                        const response = r.clone();
+                        if(response.status == 200){
+                            response.json().then((json) => {
+                                if (ajv.validate("https://image.a11y.mcgill.ca/preprocessor-response.schema.json", json)) {
+                                    SERVICE_PREPROCESSOR_MAP[preprocessor[0]] = json["name"];
+                                    // store data in cache
+                                    // disable the cache if "ca.mcgill.a11y.image.cacheTimeout" is 0
+                                    if(cacheTimeOut){
```

Comment:
```suggestion
                                    if(cacheTimeOut > 0){
```
Personal issue with using implied conversions since they're a *pain* to debug.
For example, you're not casting "cacheTimeOut" to a number above explicitly, so `if(0)` would resolve to false, but `if("0")` would resolve to true. However, both `0 > 0` and `"0" > 0` are false.